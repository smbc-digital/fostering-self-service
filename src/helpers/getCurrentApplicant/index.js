import { Applicant, SECOND_APPLICANT_ROUTE } from 'constants'

const getCurrentApplicant = match => {
    if (match.params && match.params[0] === SECOND_APPLICANT_ROUTE) {
        return Applicant.SecondApplicant
    }

    return Applicant.FirstApplicant
}

export default getCurrentApplicant