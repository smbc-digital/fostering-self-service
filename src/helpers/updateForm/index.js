import { fetchWithTimeout } from '../index'

const TaskStatus = {
    None: 0,
    Completed: 1,
    NotCompleted: 2,
    CantStart: 3
}

const StageName = {
    HomeVisit: 0,
    Application: 1
}

const HomeVisitFormName = {
    TellUsAboutYourself: 0,
    YourEmploymentDetails: 1,
    LanguagesSpokenInYourHome: 2,
    YourPartnership: 3,
    YourFosteringHistory: 4,
    YourHealth: 5,
    TellUsAboutYourInterestInFostering: 6,
    YourHousehold: 7,
    ChildrenLivingAwayFromYourHome: 8
}

const ApplicationFormName = {
    References: 0,
    GpDetails: 1,
    AddressHistory: 2
}

const getHomeVisitUpdateEndpoint = form => {
    switch (form) {
        case HomeVisitFormName.TellUsAboutYourself:
            return '/fostering/home-visit/about-yourself'
        case HomeVisitFormName.YourEmploymentDetails:
            return '/fostering/home-visit/your-employment-details'
        case HomeVisitFormName.LanguagesSpokenInYourHome:
            return '/fostering/home-visit/languages-spoken-in-your-home'
        case HomeVisitFormName.YourFosteringHistory:
            return '/fostering/home-visit/your-fostering-history'
        case HomeVisitFormName.YourPartnership:
            return '/fostering/home-visit/partnership-status'
        case HomeVisitFormName.TellUsAboutYourInterestInFostering:
            return '/fostering/home-visit/interest-in-fostering'
        case HomeVisitFormName.YourHealth:
            return '/fostering/home-visit/about-your-health'
        case HomeVisitFormName.YourHousehold:
            return '/fostering/home-visit/household'
        case HomeVisitFormName.ChildrenLivingAwayFromYourHome:
            return '/fostering/home-visit/children-living-away-from-home'
        default:
            throw new Error('No matching endpoint for given form.')
    }
}

const getApplicationUpdateEndpoint = form => {
    switch (form) {
        case ApplicationFormName.References:
            return '/fostering/application/references'
        case ApplicationFormName.GpDetails:
            return '/fostering/application/gp-details'
        default:
            throw new Error('No matching endpoint for given form.')
    }
}

const getFormStageEndpoint = stage => {
    switch (stage) {
        case StageName.HomeVisit: 
            return '/fostering/home-visit'
        case StageName.Application:
            return '/fostering/application'
        default:
            throw new Error('No matching endpoint for given stage.')
    }
}

const reduceProperties = object => Object.keys(object).reduce((acc, property) => {
    return {
        ...acc,
        [property]: object[property].value
    }
}, {})

const updateForm = async (endpoint, formData) => {
    const parsedFormData = parseFormData(formData)

    const response = await fetchWithTimeout(endpoint, {
        method: 'PATCH',
        credentials: 'include',
        body: JSON.stringify(parsedFormData),
        headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json'
        }
    }, 30000)

    if (!response.ok) {
        throw Error(response.statusText)
    }

    return await response.json()
}

const updateFormStatus = ({ form, stage, currentStatus, setStatus }) => {
    if (currentStatus === TaskStatus.None) {
        fetchWithTimeout(`${getFormStageEndpoint(stage)}/status`,
            {
                method: 'PATCH',
                credentials: 'include',
                body: JSON.stringify({
                    status: TaskStatus.NotCompleted,
                    form
                }),
                headers: {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json'
                }
            }, 30000)

        setStatus(TaskStatus.NotCompleted)
    }
}

const parseFormData = ({ firstApplicant, secondApplicant, familyReference, firstPersonalReference, secondPersonalReference, ...formData }) => {
    let parsedObject = reduceProperties(formData)

    if (firstApplicant) {
        parsedObject.firstApplicant = reduceProperties(firstApplicant)
    }

    if (secondApplicant) {
        parsedObject.secondApplicant = reduceProperties(secondApplicant)
    }

    if (familyReference) {
        parsedObject.familyReference = reduceProperties(familyReference)
        parsedObject.familyReference.address = familyReference.address.value
    }

    if (firstPersonalReference) {
        parsedObject.firstPersonalReference = reduceProperties(firstPersonalReference)
        parsedObject.firstPersonalReference.address = firstPersonalReference.address.value
    }

    if (secondPersonalReference) {
        parsedObject.secondPersonalReference = reduceProperties(secondPersonalReference)
        parsedObject.secondPersonalReference.address = secondPersonalReference.address.value
    }

    return parsedObject
}

const updateApplicationForm = async (form, formData) => {
    const endpoint = getApplicationUpdateEndpoint(form)

    return await updateForm(endpoint, formData)
}

const updateHomeVisitForm = async (form, formData) => {
    const endpoint = getHomeVisitUpdateEndpoint(form)

    return await updateForm(endpoint, formData)
}

export {
    TaskStatus,
    StageName,
    HomeVisitFormName,
    ApplicationFormName,
    updateFormStatus,
    parseFormData,
    updateApplicationForm,
    updateHomeVisitForm
}