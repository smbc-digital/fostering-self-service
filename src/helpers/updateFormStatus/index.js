export const TaskStatus = {
    None: 0,
    Completed: 1,
    NotCompleted: 2,
    CantStart: 3
}

export const FormName =
{
    TellUsAboutYourself: 0,
    YourEmploymentDetails: 1,
    LanguageSpokenInYourHome: 2,
    YourPartnership: 3,
    YourFosteringHistory: 4,
    YourHealth: 5,
    TellUsAboutYourInterestInFostering: 6,
    YourHousehold: 7,
    ChildrenLivingAwayFromYourHome: 7
}

export default function (form, currentStatus, setStatus) {
    if (currentStatus === TaskStatus.None) {
        fetch('/front-end-endpoint-update-status',
            {
                method: 'PATCH',
                credentials: 'include',
                body: JSON.stringify({
                    status: TaskStatus.NotCompleted,
                    form
                })
            })

        setStatus(TaskStatus.NotCompleted)
    }
}