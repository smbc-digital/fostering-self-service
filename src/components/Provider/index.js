import React, { useState, useEffect } from 'react'
import PropTypes from 'prop-types'
import { Context } from '../../context/'
import { fetchWithTimeout } from '../../helpers'
import { API_ROOT } from '../../config'

export const Applicant = {
	FirstApplicant: 'firstApplicant',
	secondApplicant: 'secondApplicant'
}

const reduceProperties = object => Object.keys(object).reduce((acc, property) => {
	return {
		...acc,
		[property]: {
			value: object[property] || '',
			isValid: object[property] !== null
		}
	}
}, {})

const Provider = ({ children }) => {
	const displayRecaptcha = document.getElementById('displayRecaptcha') ? document.getElementById('displayRecaptcha').innerHTML === 'true' : false
	const initialState = {
		displayRecaptcha,
		currentApplicant: Applicant.FirstApplicant,
	}

	const [isLoading, setIsLoading] = useState(true)
	const [state, setState] = useState(initialState)
	const [error, setError] = useState(undefined)

	const onChange = (event, isValid) => {
		setState({
			...state,
			[event.target.name]: {
				value: event.target.value, 
				isValid
			}})
	}

	const onChangeApplicant = (event, isValid) => {
		setState({
			...state,
			[state.currentApplicant]: {
				...state[state.currentApplicant],
				[event.target.name]: {
					value: event.target.value, 
					isValid
				}
			}
		})
	}

	const setApplicant = applicant => {
		setState({
			...state,
			currentApplicant: applicant
		})
	}

	const mapCaseToContext = caseResponse => {
		const statuses = {...caseResponse.statuses}
		let secondApplicantDetails = undefined
		delete caseResponse.statuses

		const firstApplicantDetails = reduceProperties(caseResponse.firstApplicant)
		delete caseResponse.firstApplicant

		if (caseResponse.secondApplicant !== null) {
			secondApplicantDetails = reduceProperties(caseResponse.secondApplicant)
		}
		delete caseResponse.secondApplicant

		const caseDetails = reduceProperties(caseResponse)

		setState({
			...state, 
			...statuses, 
			firstApplicant: firstApplicantDetails, 
			secondApplicant: secondApplicantDetails, 
			...caseDetails
		})
	}

	const fetchCase = async () => {
		try {
			const response = await fetchWithTimeout(`${API_ROOT}/fostering/case`, { credentials: 'include' })
			const body = await response.json()
			mapCaseToContext(body)
		} catch (error) {
			setError(error)
		}
	}

	useEffect(() => {
		fetchCase()
	}, [])

	useEffect(() => {
		// Set isLoading to false only after the case is in state 
		if (isLoading && Object.keys(state).length > Object.keys(initialState).length) {
			setIsLoading(false)
		}
	}, [state])

	if (error) {
		return <p>Error</p>
	}

	if (isLoading) {
		return <p>Loading...</p>
	}

	return <Context.Provider value={{...state, onChange, onChangeApplicant, setApplicant}}>
		{children}
	</Context.Provider>
}

Provider.propTypes = {
	children: PropTypes.object
}

export default Provider