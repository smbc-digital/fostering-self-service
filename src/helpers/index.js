import fetchWithTimeout from './fetchWithTimeout'
import { getPageRoute } from './pagehelper'
import getCurrentApplicant from './getCurrentApplicant'
import { updateFormStatus, updateForm, FormName, TaskStatus } from './updateForm'

export {
    fetchWithTimeout,
    getPageRoute,
    getCurrentApplicant,
    updateFormStatus,
    updateForm,
    FormName,
    TaskStatus
}