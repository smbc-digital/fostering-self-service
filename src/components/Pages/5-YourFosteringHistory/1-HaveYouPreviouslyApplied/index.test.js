import { React, useContextMock, renderer } from '../../../../helpers/SetupTest'
import HaveYouPreviouslyApplied from './index'
import { Applicant } from '../../../Provider'

describe('HaveYouPreviouslyApplied', () => {

    const onChangeApplicantMock = jest.fn()
    const onChangeStatusMock = jest.fn()

    beforeEach(() => {
        useContextMock.mockReturnValue({
            currentApplicant: Applicant.FirstApplicant,
            onChangeApplicant: onChangeApplicantMock,
            onChangeStatus: onChangeStatusMock,
            firstApplicant: {
                previouslyApplied: {
                    value: 'true',
                    isValid: true
                },
                firstName: {
                    value: 'first name',
                    isValid: true
                },
                lastName: {
                    value: 'last name',
                    isValid: true
                }
            }
        })
    })

    describe('snapshot', () => {
        it('renders correctly', () => {
            const match = {
                params: undefined
            }

            const tree = renderer
                .create(<HaveYouPreviouslyApplied history={{}} match={match}/>)
                .toJSON()

            expect(tree).toMatchSnapshot()
        })
    })
})