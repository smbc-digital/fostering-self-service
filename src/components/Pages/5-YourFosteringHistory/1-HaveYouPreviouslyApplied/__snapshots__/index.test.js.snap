// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HaveYouPreviouslyApplied snapshot renders correctly 1`] = `
Array [
  <h1>
    Your fostering journey
  </h1>,
  <h2>
    Your fostering history
  </h2>,
  <h2>
    first name
     
    last name
  </h2>,
  <form
    onSubmit={[Function]}
  >
    <h3
      className="no-margin-bottom"
    >
      Have you previously applied to foster, childmind or adopt?
    </h3>
    <fieldset>
      <div
        className="input-radio-container"
        role="radiogroup"
      >
        <input
          defaultChecked={true}
          disabled={undefined}
          id="have-you-previously-applied-yes"
          name="previouslyApplied"
          onChange={[Function]}
          type="radio"
          value="true"
        />
        <label
          htmlFor="have-you-previously-applied-yes"
        >
          Yes
        </label>
      </div>
      <div
        className="input-radio-container"
        role="radiogroup"
      >
        <input
          defaultChecked={false}
          disabled={undefined}
          id="have-you-previously-applied-no"
          name="previouslyApplied"
          onChange={[Function]}
          type="radio"
          value="false"
        />
        <label
          htmlFor="have-you-previously-applied-no"
        >
          No
        </label>
      </div>
    </fieldset>
    <button
      className="button-primary"
      disabled={false}
      id={undefined}
      onClick={undefined}
      type="submit"
    >
      Save and next step
    </button>
    <p>
      <a
        href="#"
        onClick={[Function]}
      >
        Back
      </a>
    </p>
    <button
      className="button-inverted"
      disabled={false}
      id={undefined}
      onClick={[Function]}
      type="submit"
    >
      Save and go back to fostering area
    </button>
  </form>,
]
`;
